visualization/dashboard/src/MLIDashboard/Cohort.ts

```diff
@@ -0,0 +1,128 @@
+    }
+
+    public getRow(index: number): {[key: string]: number} {
+        return {...this.jointDataset.dataDict[index]}
+    }
+
+    public sort(columnName: string = JointDataset.IndexLabel, reverse?: boolean): void {
+        if (this.currentSortKey !== columnName) {
+            this._filteredData.sort((a, b) => {
+                return a[columnName] - b[columnName];
+            });
+            this.currentSortKey = columnName;
+            this.currentSortReversed = false;
+        }
+        if (this.currentSortReversed !== reverse) {
+            this._filteredData.reverse();
+        }
+    }
+
+    // whether to apply bins is a decision made at the ui control level,
+    // should not mutate the true dataset. Instead, bin props are preserved
+    // and applied when requested.
+    // Bin object stores array of upper bounds for each bin, return the index
+    // if the bin of the value;
+    public unwrap(key: string, applyBin?: boolean): any[] {
+        if (applyBin && this.jointDataset.metaDict[key].isCategorical === false) {
+            let binVector = this.jointDataset.binDict[key];
+            if (binVector === undefined) {
+                this.jointDataset.addBin(key);
+                binVector = this.jointDataset.binDict[key];
+            }
+            return this._filteredData.map(row => {
+                const rowValue = row[key];
+                return binVector.findIndex(upperLimit => upperLimit >= rowValue );
+            });
+        }
+        return this._filteredData.map(row => row[key]);
+    }
+
+    public calculateAverageImportance(): number[] {
+        if (this._cachedAverageImportance) {
+            return this._cachedAverageImportance;
+        }
+
+        this._cachedAverageImportance = this.transposedLocalFeatureImportances().map(featureValues => {
+            if (!featureValues || featureValues.length === 0) {
+                return Number.NaN;
+            }
+            const total = featureValues.reduce((prev, current) => {return prev + Math.abs(current)}, 0);
+            return total / featureValues.length;
+        })
+        return this._cachedAverageImportance;
+    }
+
+    public transposedLocalFeatureImportances(): number[][] {
+        if (this._cachedTransposedLocalFeatureImportances) {
+            return this._cachedTransposedLocalFeatureImportances;
+        }
+        const featureLength = this.jointDataset.localExplanationFeatureCount;
+        const localFeatureImportances = this._filteredData.map(row => {
+            return JointDataset.localExplanationSlice(row, featureLength);
+        });
+        this._cachedTransposedLocalFeatureImportances = ModelExplanationUtils.transpose2DArray(localFeatureImportances);
+        return this._cachedTransposedLocalFeatureImportances;
+    }
+
+    private applyFilters(): void {
+        this._cachedAverageImportance = undefined;
+        this._cachedTransposedLocalFeatureImportances = undefined;
+        this.mutateCount += 1;
```

Comment:
what is mutateCount used for?  don't see it used anywhere in the code.

Reply:
used to determine if an objects internals have been changed. Currently we just take a deep copy on change, but we could optimize by checking this.